File: tests/test_statistics_export.py
#!/usr/bin/env python3
import json
import subprocess
import sys
import tempfile
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
FIXTURE = ROOT / "tests/fixtures/tui/statistics-v2.tsv"


def run(core, *args):
    return subprocess.run([core, *args], text=True, capture_output=True)


def main():
    core = sys.argv[1] if len(sys.argv) > 1 else str(ROOT / "build/hydra-core")
    result = run(core, "statistics-json", str(FIXTURE))
    assert result.returncode == 0, result.stderr
    document = json.loads(result.stdout)
    assert document["observed"] == 1788789600
    assert document["cohort"] == {"runs": 8, "steps": 11}
    assert document["metrics"]["queue"] == {
        "state": "known", "eligible": 10, "known": 7, "sum": 70,
        "mean": 10, "max": 10, "p50": 10, "p95": 10,
    }
    assert document["metrics"]["verified"]["known"] == 2
    assert document["remote"]["state"] == "unavailable"

    with tempfile.TemporaryDirectory() as folder:
        folder = Path(folder)
        malformed = folder / "malformed.tsv"
        malformed.write_text("HYDRA_STATISTICS\t2\t10\n")
        result = run(core, "statistics-json", str(malformed))
        assert result.returncode != 0 and result.stdout == ""
        partial = folder / "partial.tsv"
        partial.write_text(FIXTURE.read_text().replace("Z\t8\t11\n", ""))
        result = run(core, "statistics-json", str(partial))
        assert result.returncode != 0 and result.stdout == ""
        oversized = folder / "oversized.tsv"
        oversized.write_text("HYDRA_STATISTICS\t2\t10\nX\t" + "x" * (1024 * 1024) + "\n")
        result = run(core, "statistics-json", str(oversized))
        assert result.returncode != 0 and result.stdout == ""
        result = run(core, "statistics-compare", str(FIXTURE), str(FIXTURE))
        assert result.returncode == 0
        compared = json.loads(result.stdout)
        assert compared["delta"]["metrics"]["queue"] == {
            "state": "known", "mean": 0, "max": 0, "p50": 0, "p95": 0,
        }
    print("Statistics exporter JSON, comparison and fail-closed bounds passed")


if __name__ == "__main__":
    main()

File: tests/c/test_statistics_export.c
#include "hydra_statistics.h"
#include <assert.h>
#include <stdio.h>
#include <string.h>

static void read_all(FILE *f, char *buf, size_t cap) {
    size_t n;
    rewind(f);
    n = fread(buf, 1, cap - 1, f);
    assert(!ferror(f));
    buf[n] = '\0';
}

static bool load_fixture(struct hs_model *m) {
    FILE *f = fopen("tests/fixtures/tui/statistics-v2.tsv", "r");
    bool ok;
    assert(f);
    ok = hs_load(f, m);
    fclose(f);
    return ok;
}

int main(void) {
    struct hs_model model, other;
    struct hs_filter filter = {0};
    char json[32768];
    FILE *out;

    assert(load_fixture(&model));
    out = tmpfile(); assert(out);
    assert(hs_write_metrics_json(out, &model, &filter));
    read_all(out, json, sizeof(json));
    assert(strstr(json, "\"schema_version\":1"));
    assert(strstr(json, "\"queue\":{\"state\":\"known\""));
    assert(strstr(json, "\"eligible\":10"));
    assert(strstr(json, "\"known\":7"));
    assert(strstr(json, "\"mean\":10"));
    assert(strstr(json, "\"p95\":10"));
    fclose(out);

    filter.attention = true;
    out = tmpfile(); assert(out);
    assert(hs_write_metrics_json(out, &model, &filter));
    read_all(out, json, sizeof(json));
    assert(strstr(json, "\"recoveries\":{\"state\":\"known\""));
    fclose(out);

    assert(load_fixture(&other));
    other.runs[0].started = 0;
    out = tmpfile(); assert(out);
    assert(hs_write_metrics_compare_json(out, &model, &other, &filter));
    read_all(out, json, sizeof(json));
    assert(strstr(json, "\"delta\":{\"metrics\""));
    assert(strstr(json, "\"elapsed\":{\"state\":\"known\""));
    fclose(out);

    /* A malformed feed never reaches the exporter: callers must fail closed. */
    out = tmpfile(); assert(out);
    assert(!hs_load(out, &other));
    fclose(out);
    puts("Statistics exporter: bounded JSON metrics, comparison and fail-closed input passed");
    return 0;
}
